### DISCUSSION TITLE
Extracting Text from the casts - or - pure text casts without terminal control codes (for written documentation)

### DISCUSSION THREAD

**Discussion by klartext** [asciinema/discussions#45] [Ideas]
For documentation of "things I did in the terminal" I want to use asciinema for recording, and then extract the text parts,
so that these parts can then be imported into textfiles, markdown files or LaTeX files (as verbatim text).

Would be nice if asciinema could offer a "pure text" output format - without the terminal control codes.

I ask, if it would be possible to add this as an option to asciinema.
(some kind of 'raw'  mode, just without the terminal symbls, so maybe 'rawtext' mode)

**Comment by bkil** [comment]
Yes. It could even add light markdown (LaTeX, HTML, ...) formatting as well to differentiate between what you type and what appears on the screen and such.

**Comment by klartext** [comment]
When asking if it could be done, I meant: will there be the decision from the asciinema developers to add that feature and will it be added then. (And when will that happen?)

But recently (two days ago) I found the documentation on the output format
https://github.com/asciinema/asciinema/blob/develop/doc/asciicast-v2.md
and so I started with writing my own tool for that task.

Maybe a seperate tool better matches the unix philosophy...
...but so far the code of the tool is only residing on my laptop.
So, I will let the issue stay open.

**Comment by bkil** [comment]
Related: https://github.com/asciinema/asciinema-server/issues/111 https://github.com/asciinema/asciinema-player/issues/156

**Comment by klartext** [comment]
Listen to what Marcin Kulik talks about here:

https://changelog.com/podcast/561

Listen at 1:03:37.

That's what I want: the exportable "freezed" textversion of the terminal output.

**Comment by ku1ik** [comment]
In order to reliably convert captured recording to pure text it needs to run through some kind of terminal emulator and the text without color attributes needs to be extracted from terminal's screen buffer. All other approaches, like stripping certain sequences with regex etc, are brittle and give good results only in simple cases.

[avt](https://github.com/asciinema/avt), used by asciinema player, server and [agg](https://github.com/asciinema/agg), can do this. This is Rust library which is embeded in the mentioned components, but it's not embedded in `asciinema` CLI. The CLI is written in Python, and in order to use avt in it we'd need to use something like [PyO3](https://github.com/PyO3/pyo3), but that requires whole Rust build toolchain, which complicates building and packaging of the CLI, so I'm hesitatint here. 

But avt can be invoked separately to do to-text conversion like this (in avt checkout, assuming you have Rust toolchain installed):

```
$ cd avt
$ printf '\e[1mhe\e[41mll\r\no!' | cargo run -q --example text
hell
o!
```

The above one converted arbitrary text with sequences in it. One below converts from .cast file:

```
asciinema cat demo.cast | cargo run -q --example text
```

See demo here:

[![asciicast](https://asciinema.org/a/614835.svg)](https://asciinema.org/a/614835)

Ideally it would be something built in like, `asciinema cat --text demo.cast` or similar, I know :) I'm considering Rust rewrite of the CLI (for several other reasons), so if that happens then it would be trivial to have built-in first-class support for this use case in the CLI.

**Comment by ku1ik** [comment]
As of today (via [this commit](https://github.com/asciinema/asciinema-server/commit/5e95167d3b265996072485280420f0c68a8ab24b)) you can download pure text version of the recording from asciinema.org by appending `.txt` to recording page URL.

For example, for the demo above which has URL https://asciinema.org/a/614835 you can download `.txt` from https://asciinema.org/a/614835.txt

Need to add some link/button in the UI for easier discovery of this, but the endpoint is fully functional now :sparkles:

**Reply by klartext** [reply]
... that's nice, but I'm waiting for this feature to be available as offline version,
so that I can do anything locally.

If I understood correctly what you have written somewhere above, I could do it with the Rust environment installed?
So far I have not found avt and agg installable as binary via package manager.
I did not use Rust so far, so I'm not familiar with that environment.

**Reply by ku1ik** [reply]
Got it. Yes, you can do it locally with avt, but it requires building avt locally with Rust toolchain. I'll write instructions on how to do that when I have moment.

**Comment by ku1ik** [comment]
And it's now available via Download dropdown:


![Screenshot from 2023-10-17 17-45-23](https://github.com/asciinema/discussions/assets/17589/b544703e-0e18-4ad4-849d-d4692c8998b2)

**Comment by tennox** [comment]
I stumbled here just to find out now there is:
```console
asciinema cat foo.cast
```

**Reply by ku1ik** [reply]
Note that `asciinema cat <filename>` doesn't output plain text, but rather the raw capture output, which typically includes control sequences (colors, cursor movement, etc).

**Reply by LeonardoMor** [reply]
Yes. You can see the control characters by redirecting that command to a text file and then reading that.

**Comment by ku1ik** [comment]
Good news: the upcoming asciinema CLI 3.0 supports recording and conversion to plain text files.

Record to plain text file:

```sh
asciinema rec foo.txt
asciinema rec foo.log --format txt
```

Convert existing cast file to plain text:

```sh
asciinema convert foo.cast foo.txt
asciinema convert foo.cast foo.log --format txt
```

If you'd like to play with it then check here: https://github.com/asciinema/asciinema/releases/tag/v3.0.0-rc.1

You can either download a precompiled binary from the above page or [build from source](https://discourse.asciinema.org/t/testing-the-new-rust-version-of-the-asciinema-cli/778).

**Comment by ku1ik** [comment]
Since asciinema CLI 3.0 there's ability to export as plain text, with `asciinema convert foo.cast foo.txt`, as well as raw output (including control sequences), with `asciinema convert foo.cast foo.raw -f raw`.

**Comment by klartext** [comment]
Thank you for implementing this!
